### ISSUE TITLE
No GitHub Actions should run on PRs from forks

### ISSUE THREAD

**Issue by jessebrennan** [DataBiosphere/azul#2984]
... or at least the definition of the action on the `develop` branch should be used.

Essentially, we should not allow code from external collaborators to be run within out security domain.

**Comment by theathorn** [comment]
@jessebrennan to provide reproduction steps.

**Comment by hannes-ucsc** [comment]
Case in point: secret leak in CI process:

![image](https://user-images.githubusercontent.com/5143256/114907602-1a018e00-9dd0-11eb-9edf-b2b6669dfc70.png)

**Comment by jessebrennan** [comment]
There are a couple ways we can address this. First, here's how things currently work. For PRs from forks, Github restricts the token it provides to the Action to be "read only". If there are any secrets provided to the actions, they are withheld. My guess is that, since Github is giving away their compute for free to anyone who uses their service, avoiding abuse in cases like this isn't really a concern. For self-hosted runners, [Github suggests not even to connect them to public repositories](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories).

Github doesn't directly provide a way to prevent *any* actions from running when a PR is from a fork if the repository is public (it's [only possible for private repos](https://docs.github.com/en/github/administering-a-repository/disabling-or-limiting-github-actions-for-a-repository#enabling-workflows-for-private-repository-forks)).  Here's what we *can* do:

- [Control which actions are allowed to run](https://docs.github.com/en/github/administering-a-repository/disabling-or-limiting-github-actions-for-a-repository#managing-github-actions-permissions-for-your-repository). This prevents us from using any "premade" Github action we find out in the wild in the Actions Marketplace for example. We can restrict so that only actions within The DataBiosphere org are run in Azul.

  Unfortunately, even with this restriction, an action modified by a PR from a fork can still run arbitrary code, so this doesn't address the root concern.

- Completely disable forks from Azul. This is the only way I have found or can think of to actually prevent PRs from forks from running actions.

In conclusion, we can either live with the fact that PRs from forks may run whatever code they want in Github Actions, or we can restrict it so that one cannot make forks from Azul.

**Comment by jessebrennan** [comment]
@hannes-ucsc and I discussed this and decided that since we cannot restrict actions that run on forks, we will move this issue to triage and likely close it.

In #2768 we will include an action item to control which actions are allowed to run, setting a permission boundary so that only the actions we explicitly need are allowed.

I will also investigate what the potential damage is that can be caused by a PR from a fork given, even with Github's restricted token.

**Comment by jessebrennan** [comment]
Potential damage:

- We could have to manually close some malicious pull requests like this https://dev.to/thibaultduponchelle/the-github-action-mining-attack-through-pull-request-2lmc
- Public data about our repository could be read or harvested

I honestly can't think of anything beyond that at this point. Since the Github token for fork PRs is read only, they cannot make any changes to our repo. Secrets aren't provided to these types of PRs, so they wouldn't be able to affect Coveralls or Codecov. Compute would be wasted, but that's just Github's money, not ours.

**Comment by jessebrennan** [comment]
@hannes-ucsc: "Jesse to determine why the attack needs a fork and a PR against the upstream repository. Who 'pays' for the actions run, the fork or the upstream repository? Are we at risk of being denied access to Github Actions. Ask Github support these questions."

**Comment by jessebrennan** [comment]
Github has taken steps to resolve this now. For one, upstream repositories won't be penalized for abusive actions run on fork PRs.

> Specifically, if we determine an Actions run to be abusive or against our terms, our enforcement will be directed at the account hosting the fork and not the account associated with the upstream repository.

Github now requires approval from a maintainer before actions are run on PRs from forks if the PR is from a **first time contriutor**.

> Second, pull requests from first-time contributors will require manual approval from a repository collaborator with write access before any Actions workflows run.

See https://github.blog/2021-04-22-github-actions-update-helping-maintainers-combat-bad-actors/ for details.

I think it's safe to close this now.
